workflow/artifacts/s3/s3.go

```diff
@@ -180,7 +180,27 @@ func (s3Driver *ArtifactDriver) Delete(artifact *wfv1.Artifact) error {
 		if err != nil {
 			return err
 		}
-		return s3cli.Delete(artifact.S3.Bucket, artifact.S3.Key)
+
+		isDir, err := s3cli.IsDirectory(artifact.S3.Bucket, artifact.S3.Key)
+		if err != nil {
+			return fmt.Errorf("failed to test if %s is a directory: %v", artifact.S3.Key, err)
+		}
+
+		if !isDir {
+			return s3cli.Delete(artifact.S3.Bucket, artifact.S3.Key)
+		} else {
```

Comment:
this could potentially be optimized I'm thinking -- try deleting and if it doesn't work and gives the appropriate error, then proceed with the directory deletion

similar to what I did in #12974

Reply:
Make sense, the current change doubles the API calls

Reply:
However, looking deeper I realise
For 
```
err = deleteFile
if isNotFound(err) {
    deleteDir
}
```
to work, `deleteFile` needs to return the appropriate error.
In s3, delete only adds a marker and will return `204` regardless if the key exist or not
- https://stackoverflow.com/questions/30697746/why-does-s3-deleteobject-not-fail-when-the-specified-key-doesnt-exist
- https://github.com/minio/minio-go/blob/master/api-remove.go#L198-L200

Reply:
Updated, use simple path suffix check instead of s3cli.ListDirectory to determine if the path is a directory

Reply:
Huh does the `/` check actually work? Great workaround if so!